Cache the whole installation when possible

lazka opened the pull request:
In case release=true we try to cache the whole installation.

As cache restore key we use the whole input + the installer checksum, so
we only get a similar setup back.

For saving we hash the content of /var/lib/pacman/local which contains all the info
about which packages are installed.

Fixes #50

eine commented:
Looks good! However, it feels that PackageCache is a superset of InstallCache (restore and save are almost the same). Is it possible to define a base class, and then extend others from there?

BTW, if you find it easier to handle, those classes can be defined in a different file.

lazka commented:
> Is it possible to define a base class, and then extend others from there?

I'd rather not tbh, as I think it would be harder to understand if I split it up. I've moved some common things out of save though now (saveCacheMaybe, hashPath)

lazka commented:
So here are runs from #50:

* saving cache 8m40s: https://github.com/lazka/setup-msys2/runs/904253520
* restore 2m26s (1278 MB): https://github.com/lazka/setup-msys2/runs/904284561
* no cache 4m6s: https://github.com/ComFreek/curv/runs/886676572

We could save one minute with the first run by not saving the package cache.

eine commented:
> I'd rather not tbh, as I think it would be harder to understand if I split it up. I've moved some common things out of save though now (saveCacheMaybe, hashPath)

What about defining `restore(paths, key, restoreKeys)` and using it as `restore([this.pkgCachePath], this.jobCacheKey, [this.jobCacheKey, this.fallbackCacheKey])` from `PackageCache`; and `restore([this.msysRootDir], this.jobCacheKey, [this.jobCacheKey])` from `InstallCache`? The same for L99-L107 and L142-L150.

> So, in case the config stays the same it's worth it. If you experiment and change config a lot it might be a bit annoying. And if you hit the 5GB limit on each run it will be super slow.

I think this is expected. Users should be aware that changing the config frequently will produce multiple cache misses. By the same token, users should be aware of the limits. The limit does not only apply to caching; some packages cannot be built because of RAM or disk space. The use case we want to accelerate is users running the same config very frequently.

I believe that the time savings for frequent runs are worth having overhead the first 1-2 runs after each change. Not so long ago we were having +12min setup time. Now we are talking about 1-2min. That's awesome.

lazka commented:
> What about defining `restore(paths, key, restoreKeys)` and using it as `restore([this.pkgCachePath], this.jobCacheKey, [this.jobCacheKey, this.fallbackCacheKey])` from `PackageCache`; and `restore([this.msysRootDir], this.jobCacheKey, [this.jobCacheKey])` from `InstallCache`? The same for L99-L107 and L142-L150.

ok, done.

> I think this is expected. Users should be aware that changing the config frequently will produce multiple cache misses. By the same token, users should be aware of the limits. The limit does not only apply to caching; some packages cannot be built because of RAM or disk space. The use case we want to accelerate is users running the same config very frequently.

yeah. if this really is a problem they can move the package installation into the next run step so the packages don't get cached.

----

I already told you in chat, but if https://github.com/actions/cache/issues/301 gets resolved we'll get faster cache saving as well.

lazka commented:
One remaining question: If we know we are going to cache the whole installation, should we bother saving the package cache?

eine commented on main.js:
```diff
@@ -140,26 +191,36 @@ async function run() {
       core.endGroup();
+
+      if (!cachedInstall) {
+        // Use upstream package instead of the default installation in the virtual environment.
```
This comment should be above, before or after `if (input.release)`.

eine commented:
Thanks! :tada:
